developing/vcs/README.md

```diff
@@ -2,7 +2,321 @@
+
+## Repo maintenance and hygiene
+
+* Avoid including files in source control that are specific to your development machine or process.
+* Ignore files that are specific to your machine or process in your
+`~/.gitignore_global` as opposed to your repo's `.gitignore`.
+* Delete local and remote feature branches after merging. This can be automated
+on the GitHub side with [Probot](https://probot.github.io/apps/delete-merged-branch/).
+Locally, you can add an alias to your `~/.gitconfig` file. Example:
+
+  ```
+  [alias]
+    fresh = "!git co master && git pull && git branch --merged | grep -v '\\*' | xargs -n 1 git branch -d"
+  ```
+
+  Then instead of doing the typical `git checkout master && git pull`, you would
+  only run `git fresh`.
+
+### GitHub repo settings
+
+* [Protect the master branch](https://help.github.com/en/articles/configuring-protected-branches)
+by turning on the following settings at a minimum:
+
+  * Require pull request reviews before merging
+  * Require status checks to pass before merging
+  * Include administrators
+
+* Under `Settings->Merge button`:
+  * Uncheck all options except for `Allow merge commits` because it's the only
+  one that allows deleting local branches with `-d`. We should never have to use
+  `-D` to delete a branch.
+
+## PR size
+
+Pull requests should be small and focused enough to be reviewable in under 15,
+or ideally under 5, minutes.
+
+## Git workflow
+
+### Fork vs clone
+
+If it's a repo you have write and push access to, clone it. Otherwise, fork it.
+
+### Cloning with SSH vs HTTPS
+
+Do we have strong opinions on this?
+
+### Merge vs rebase
+
+#### TL;DR
+
+If you care about keeping a clean commit history, where each commit has a great
+message, represents a meaningful change, and passes tests and other checks, you
+should use `rebase`. A history that is easy to read and follow is more valuable
+than one that is hard to understand. The full history remains available via
+`git reflog`.
+
+The closest we can come to the best of both worlds is to use `merge` locally,
+and then squash commits via the GitHub website. This comes with two downsides
+though: we are forced to use `-D` to delete local branches, and we become
+dependent on GitHub.
+
+#### Merge
+
+Pros:
+
+* Safer since you don't have to force push
+* History does not change, and includes the good, bad, and ugly
+
+Cons:
+
+* Noisier history with commits that don't add any value and that
+can cause confusion when reviewing the log
+* Not all commits might pass tests, making debugging hard
+* Cannot use interactive rebase to squash and/or reorder commits
+* Increases risk of losing code when handling conflicts due to working with
```

Comment:
Are we sure about this one? Aren’t conflicts about the current state of the source and destination branches and don’t involve the history at all?

Reply:
This was poorly worded by me and will be removed. See discussion above started by MacRae. Also, most of this will be rewritten. It was a first draft to start the discussion.